alf/algorithms/algorithm.py

```diff
@@ -187,52 +205,46 @@ def _get_trainable_vars(obj):
         module_sets = [copy.copy(s) for s in self._init_module_sets]
         optimizers = copy.copy(self._init_optimizers)
         algorithm_names = [self.name] * len(optimizers)
+        init_module_ids = set(map(id, sum(module_sets, [])))
 
         # This set stores all the seen distinct variables so far in this alg
-        var_ids = set(
-            map(
-                id,
-                tf.nest.flatten(
-                    tf.nest.map_structure(_get_trainable_vars, module_sets))))
+        vars = tf.nest.flatten(
+            tf.nest.map_structure(_get_trainable_vars, module_sets))
+        var_ids = set(map(id, vars))
+
+        def _check_vars(new_vars):
+            new_var_ids = set(map(id, new_vars))
+            dup_ids = var_ids & new_var_ids
+            dup_vars = [v for v in new_vars if id(v) in dup_ids]
+            assert not dup_vars, \
```

Comment:
It's better to report the error on module level if it is available: def _check_trainable(var_or_module)